osutil/epoll/epoll.go

```diff
@@ -0,0 +1,201 @@
+
+// RegisteredFdCount returns the number of file descriptors which are currently
+// registered to the epoll instance.
+func (e *Epoll) RegisteredFdCount() int {
+	return int(atomic.LoadInt32(&e.registeredFdCount))
+}
+
+func (e *Epoll) incrementRegisteredFdCount() {
+	atomic.AddInt32(&e.registeredFdCount, 1)
+}
+
+func (e *Epoll) decrementRegisteredFdCount() {
+	atomic.AddInt32(&e.registeredFdCount, -1)
+}
+
+func (e *Epoll) zeroRegisteredFdCount() {
+	atomic.StoreInt32(&e.registeredFdCount, 0)
+}
+
+// Register registers a file descriptor and allows observing speicifc I/O readiness events.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_ADD for details.
+func (e *Epoll) Register(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_ADD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	if err != nil {
+		return err
+	}
+	e.incrementRegisteredFdCount()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Deregister removes the given file descriptor from the epoll instance.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_DEL for details.
+func (e *Epoll) Deregister(fd int) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_DEL, fd, &unix.EpollEvent{})
+	if err != nil {
+		return err
+	}
+	e.decrementRegisteredFdCount()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Modify changes the set of monitored I/O readiness events of a previously registered file descriptor.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_MOD for details.
+func (e *Epoll) Modify(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_MOD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Event describes an IO readiness event on a specific file descriptor.
+type Event struct {
+	Fd        int
+	Readiness Readiness
+}
+
+// WaitTimeout blocks and waits with the given timeout for arrival of events on any of the added file descriptors.
+//
+// A msec value of -1 disables timeout.
+//
+// Please refer to epoll_wait(2) and EPOLL_WAIT for details.
+//
+// Warning, using epoll from Golang explicitly is tricky.
+func (e *Epoll) WaitTimeout(duration time.Duration) ([]Event, error) {
+	msec := int(duration.Milliseconds())
+	if duration < 0 {
+		msec = -1
+	}
+	n := 0
+	var err error
+	var sysEvents []unix.EpollEvent
+	for {
+		bufLen := e.RegisteredFdCount()
+		if bufLen < 1 {
+			bufLen = 1
+		}
+		sysEvents = make([]unix.EpollEvent, bufLen)
+		startTs := time.Now()
+		n, err = unix.EpollWait(e.fd, sysEvents, msec)
+		runtime.KeepAlive(e)
+		// unix.EpollWait can return unix.EINTR, which we want to handle by
+		// adjusting the timeout (if necessary) and restarting the syscall
+		if err == nil {
+			break
+		} else if err != unix.EINTR {
+			return nil, err
+		}
+		if msec == -1 {
+			continue
+		}
+		elapsed := time.Since(startTs)
+		msec -= int(elapsed.Milliseconds())
+		if msec <= 0 {
+			n = 0
+			break
+		}
+	}
+	events := make([]Event, 0, n)
```

Comment:
here we could have a shortcut:

if n == 0 {
    return nil, nil
}

Reply:
I considered this, but I think we want to allow calling `Wait()` when no FDs are registered yet, for two reasons:

1. The epoll_wait syscall allows waiting on an epoll FD when no other FDs are yet registered to be listened
2. There's no way to interrupt the epoll_syscall if the final registered FD is deregistered while the call is still waiting

I would like the behavior to be consistent between `Wait()` with no registered FDs and `Wait()` during which the final FD is deregistered, and I think it's reasonable for both to match the behavior of the epoll_wait syscall.  I discussed this a bit more in the commit message https://github.com/snapcore/snapd/pull/12963/commits/1342f9edab7235b708eea57a2b3f69126ffc6b45

Please let me know what you think about this. If you want to error out when there are no registered FDs, do you have any ideas for how to interrupt the epoll_wait syscall if/when the final FD is deregistered?

Reply:
I'm a bit confused by your comment, as what I'm saying is about the return value, not the call to Wait

Reply:
Oh, yes, I'm sorry, I read `events` as `sysEvents`. You're right, an early return is a good idea, thanks!